homeassistant/components/evohome/coordinator.py

```diff
@@ -5,85 +5,94 @@
+
+from .const import CONF_LOCATION_IDX, DOMAIN, GWS, TCS, UTC_OFFSET
+from .helpers import handle_evo_exception
+
+if TYPE_CHECKING:
+    from . import EvoSession
 
 _LOGGER = logging.getLogger(__name__.rpartition(".")[0])
 
 
 class EvoBroker:
-    """Container for evohome client and data."""
+    """Broker for evohome client broker."""
+
+    def __init__(self, sess: EvoSession) -> None:
+        """Initialize the evohome broker and its data structure."""
+
+        self._sess = sess
+        self.hass = sess.hass
+
+        assert isinstance(sess.client_v2, evo.EvohomeClient)  # mypy
+
+        self.client = sess.client_v2
+        self.client_v1 = sess.client_v1
+
+        self.loc_idx: int = None  # type: ignore[assignment]
+        self.loc: evo.Location = None  # type: ignore[assignment]
+
+        self.loc_utc_offset: timedelta = None  # type: ignore[assignment]
+        self.tcs: evo.ControlSystem = None  # type: ignore[assignment]
 
-    def __init__(
-        self,
-        hass: HomeAssistant,
-        client: evo.EvohomeClient,
-        client_v1: ev1.EvohomeClient | None,
-        store: Store[dict[str, Any]],
-        params: ConfigType,
-    ) -> None:
-        """Initialize the evohome client and its data structure."""
-        self.hass = hass
-        self.client = client
-        self.client_v1 = client_v1
-        self._store = store
-        self.params = params
-
-        loc_idx = params[CONF_LOCATION_IDX]
-        self._location: evo.Location = client.locations[loc_idx]
-
-        assert isinstance(client.installation_info, list)  # mypy
-
-        self.config = client.installation_info[loc_idx][GWS][0][TCS][0]
-        self.tcs: evo.ControlSystem = self._location._gateways[0]._control_systems[0]  # noqa: SLF001
-        self.loc_utc_offset = timedelta(minutes=self._location.timeZone[UTC_OFFSET])
         self.temps: dict[str, float | None] = {}
 
-    async def save_auth_tokens(self) -> None:
-        """Save access tokens and session IDs to the store for later use."""
-        # evohomeasync2 uses naive/local datetimes
-        access_token_expires = dt_local_to_aware(
-            self.client.access_token_expires  # type: ignore[arg-type]
-        )
+    def validate_location(self, loc_idx: int) -> bool:
+        """Get the default TCS of the specified location."""
 
-        app_storage: dict[str, Any] = {
-            CONF_USERNAME: self.client.username,
-            REFRESH_TOKEN: self.client.refresh_token,
-            ACCESS_TOKEN: self.client.access_token,
-            ACCESS_TOKEN_EXPIRES: access_token_expires.isoformat(),
-        }
+        self.loc_idx = loc_idx
 
-        if self.client_v1:
-            app_storage[USER_DATA] = {
-                SZ_SESSION_ID: self.client_v1.broker.session_id,
-            }  # this is the schema for STORAGE_VER == 1
-        else:
-            app_storage[USER_DATA] = {}
+        assert isinstance(self.client.installation_info, list)  # mypy
```

Comment:
This signals to me that there are design decisions that doesn't fit with you how want to do things really as it shouldn't be necessary

Reply:
I am not sure what you mean...

When the client is instantiated, it is `None`, after calling `EvoSession.Authenticate() --> aiohttp.GET()`, it is a list. 

Since the dependency (`EvoSession`) is injected into `EvoBroker`, it will be a list - it's just that **mypy** won't know that - hence the assert as a mypy hint.

It's a similar issue to: 
```python
self.loc_idx: int = None  # type: ignore[assignment]
```
... vs:
```
self.loc_idx: int | None = None
```

Reply:
question: is there a design pattern for this that I could use instead - perhaps expose underlying attrs in the via a `@property` in the dependant class?

Reply:
Once again its a design thing, you're in control of the library but you get into situations where you have to work around your own design.

At least skip isinstance as that is preferably avoided due to it is a slow process.
Maybe flip it around to be 
```suggestion
        assert self.client.installation_info is not None
```